Tools/cases_generator/tier2_abstract_generator.py

```diff
@@ -0,0 +1,252 @@
+
+
+def decref_inputs(
+    out: CWriter,
+    tkn: Token,
+    tkn_iter: Iterator[Token],
+    uop: Uop,
+    stack: Stack,
+    inst: Instruction | None,
+) -> None:
+    next(tkn_iter)
+    next(tkn_iter)
+    next(tkn_iter)
+    out.emit_at("", tkn)
+
+
+def emit_default(out: CWriter, uop: Uop) -> None:
+    for i, var in enumerate(uop.stack.outputs):
+        if var.name != "unused" and not var.peek:
+            if var.is_array():
+                out.emit(f"for (int _i = {var.size}; --_i >= 0;) {{\n")
+                out.emit(f"{var.name}[_i] = sym_init_unknown(ctx);\n")
+                out.emit(f"if ({var.name}[_i] == NULL) goto error;\n")
+                out.emit("}\n")
+            elif var.name == "null":
+                out.emit(f"{var.name} = sym_init_unknown(ctx);\n")
+                out.emit(f"if ({var.name} == NULL) goto error;\n")
+            else:
+                out.emit(f"{var.name} = sym_init_unknown(ctx);\n")
+                out.emit(f"if ({var.name} == NULL) goto error;\n")
+
+
+def write_uop(
+    override: Uop | None,
+    uop: Uop,
+    out: CWriter,
+    stack: Stack,
+    debug: bool,
+    skip_inputs: bool,
+) -> None:
+    try:
+        has_type_prop = any(
+            output.type_prop is not None for output in (override or uop).stack.outputs
+        )
+        prototype = override if override else uop
+        is_override = override is not None
+        out.start_line()
+        for var in reversed(prototype.stack.inputs):
+            if not skip_inputs or (has_type_prop and var.peek):
+                out.emit(stack.pop(var))
+        if not prototype.properties.stores_sp:
+            for i, var in enumerate(prototype.stack.outputs):
+                if not var.peek or is_override or has_type_prop:
+                    out.emit(stack.push(var))
+        if debug:
+            args = []
+            for var in prototype.stack.inputs:
+                if not var.peek or is_override:
+                    args.append(var.name)
+            out.emit(f'DEBUG_PRINTF({", ".join(args)});\n')
+        if override or has_type_prop:
+            for cache in uop.caches:
+                if cache.name != "unused":
+                    if cache.size == 4:
+                        type = cast = "PyObject *"
+                    else:
+                        type = f"uint{cache.size*16}_t "
+                        cast = f"uint{cache.size*16}_t"
+                    out.emit(f"{type}{cache.name} = ({cast})inst->operand;\n")
+        if override:
+            replacement_funcs = {
+                "DECREF_INPUTS": decref_inputs,
+                "SYNC_SP": replace_sync_sp,
+            }
+            emit_tokens(out, override, stack, None, replacement_funcs)
+        else:
+            emit_default(out, uop)
+        # Type propagation
+        if has_type_prop:
+            out.emit("\n")
+        for output in (override or uop).stack.outputs:
+            if typ := output.type_prop:
```

Comment:
This needs to go, I'm afraid.
Doing this here prevents us using the abstract interpreter generator from generating abstract interpreters, and limits us to generating a type propagation pass.

Not only that, but it cannot handle anything more complex, like the various `TO_BOOL` specializations, so they need to be written out manually.

We will almost certainly want to reuse the generator for other passes, and this prevents that.